README.md

```diff
@@ -188,8 +191,59 @@ groups:
 
 This could be used like so: `crane provision service1`, `crane run -v databases` or `crane lift -r services database1`. `crane status` is an alias for `crane status default`, which in that example is an alias for `crane status service1 database1`.
 
+### Cascading commands
+
 When using targets, it is also possible to cascade the commands to related containers. There are 2 different flags, `--cascade-affected` and `--cascade-dependencies`. In our example configuration above, when targeting the `mysql` container, the `apache` container would be considered to be "affected". When targeting the `apache` container, the `mysql` container would be considered as a "dependency". Both flags take a string argument, which specifies which type of cascading is desired, options are `volumesFrom`, `link`, `net` and `all`.
 
+### Hooks
+
+In order to run certain commands before or after key lifecycle events of containers, hooks can declared in the configuration. They are run synchronously on the host where Crane is installed, outside containers, via an `exec` call. They may interrupt the flow by returning a non-zero status. If shell features more advanced than basic variable expansion is required, you should explicitly spawn a shell to run the command in (`sh -c 'ls *'`).
+
+Hooks are declared at the top level of the configuration, under the `hooks` key. See YAML example below:
+
+```
+containers:
+  service1:
+    image: busybox
+    run:
+      detach: true
+      cmd: ["sleep", "50"]
+  service2:
+    image: busybox
+    run:
+      detach: true
+      cmd: ["sleep", "50"]
+  service3:
+    image: busybox
+    run:
+      detach: true
+      cmd: ["sleep", "50"]
+groups:
+  foo:
+    - service1
+    - service2
+  bar:
+    - service2
+    - service3
+hooks:
+  foo:
+    post-start: echo container from foo started
+  bar:
+    post-stop: echo container from bar stopped
+  service3:
+    post-stop: echo container service3 stopped
+```
+
+Hooks can be defined on a group level (`foo`, `bar`) so that they apply to all containers within that group, or directly on a container (`service3`). At most one hook can be registered per container and per event. When more than one hook is found for a given container and a given event, the following rules apply:
```

Comment:
I restricted the usage of several hooks compared to what was discussed in #148, to keep implementation concise and avoid non-deterministic behavior.